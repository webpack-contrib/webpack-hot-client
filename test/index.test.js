/* eslint-disable global-require */

const webpack = require('webpack');
const WebSocket = require('ws');

const client = require('../lib');

const logLevel = 'silent';
const options = { logLevel };
const validTypes = ['compile', 'hash', 'ok', 'warnings'];

describe('api', () => {
  test('array config', (done) => {
    const config = require('./fixtures/webpack.config-array.js');
    const compiler = webpack(config);
    const { server } = client(compiler, options);

    server.on('listening', () => {
      const { host, port } = server;
      const socket = new WebSocket(`ws://${host}:${port}`);

      socket.on('message', (raw) => {
        const data = JSON.parse(raw);

        if (data.type === 'errors') {
          console.log(data); // eslint-disable-line no-console
        }

        expect(validTypes).toContain(data.type);

        if (data.type === 'hash') {
          server.close(done);
        }
      });

      socket.on('open', () => {
        compiler.run(() => {});
      });
    });
  });

  test('object config', (done) => {
    const config = require('./fixtures/webpack.config-object.js');
    const compiler = webpack(config);
    const { server } = client(compiler, options);

    server.on('listening', () => {
      const { host, port } = server;
      const socket = new WebSocket(`ws://${host}:${port}`);

      socket.on('message', (raw) => {
        const data = JSON.parse(raw);

        expect(validTypes).toContain(data.type);

        if (data.type === 'hash') {
          server.close(done);
        }
      });

      socket.on('open', () => {
        compiler.run(() => {});
      });
    });
  });

  test('function returns array', (done) => {
    const config = require('./fixtures/webpack.config-function.js');
    const compiler = webpack(config);
    const { server } = client(compiler, options);

    server.on('listening', () => {
      const { host, port } = server;
      const socket = new WebSocket(`ws://${host}:${port}`);

      socket.on('message', (raw) => {
        const data = JSON.parse(raw);

        expect(validTypes).toContain(data.type);

        if (data.type === 'hash') {
          server.close(done);
        }
      });

      socket.on('open', () => {
        compiler.run(() => {});
      });
    });
  });

  test('MultiCompiler config', (done) => {
    const config = require('./fixtures/multi/webpack.config.js');
    const compiler = webpack(config);
    const { server } = client(compiler, options);

    server.on('listening', () => {
      const { host, port } = server;
      const socket = new WebSocket(`ws://${host}:${port}`);

      socket.on('message', (raw) => {
        const data = JSON.parse(raw);

        expect(validTypes).toContain(data.type);

        if (data.type === 'hash') {
          server.close(done);
        }
      });

      socket.on('open', () => {
        compiler.run(() => {});
      });
    });
  });

  test('options sanity check', (done) => {
    const config = require('./fixtures/webpack.config-object.js');
    const compiler = webpack(config);
    const { options: opts, server } = client(compiler, options);

    server.on('listening', () => {
      const { host, port } = server;
      expect({ host, port }).toMatchSnapshot();
      expect(opts).toMatchSnapshot();

      setTimeout(() => server.close(done), 500);
    });
  });

  test('mismatched client/server options sanity check', (done) => {
    const config = require('./fixtures/webpack.config-object.js');
    const compiler = webpack(config);
    const clientOptions = Object.assign({}, options, {
      host: {
        client: 'localhost',
        server: '0.0.0.0',
      },
      port: {
        client: 6000,
        server: 7000,
      },
    });
    const { options: opts, server } = client(compiler, clientOptions);

    server.on('listening', () => {
      const { host, port } = server;
      expect({ host, port }).toMatchSnapshot();
      expect(opts).toMatchSnapshot();

      setTimeout(() => server.close(done), 500);
    });
  });
});
