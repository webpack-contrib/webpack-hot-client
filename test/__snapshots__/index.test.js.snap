// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api mismatched client/server options sanity check 1`] = `
Object {
  "host": "0.0.0.0",
  "port": 7000,
}
`;

exports[`api mismatched client/server options sanity check 2`] = `
Object {
  "allEntries": false,
  "autoConfigure": true,
  "hmr": true,
  "host": Object {
    "client": "localhost",
    "server": "0.0.0.0",
  },
  "https": undefined,
  "log": LogLevel {
    "currentLevel": 5,
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "log": [Function],
    "methodFactory": PrefixFactory {
      "options": Object {
        "level": [Function],
        "name": [Function],
        "template": "{{level}} [90mï½¢{{name}}ï½£[39m: ",
        "time": [Function],
      },
      Symbol(a log instance): [Circular],
      Symbol(valid log levels): Object {
        "DEBUG": 1,
        "ERROR": 4,
        "INFO": 2,
        "SILENT": 5,
        "TRACE": 0,
        "WARN": 3,
      },
    },
    "name": "hot",
    "options": Object {
      "factory": null,
      "level": "silent",
      "name": "hot",
      "prefix": Object {
        "level": [Function],
        "template": "{{level}} [90mï½¢{{name}}ï½£[39m: ",
      },
      "timestamp": false,
      "unique": true,
    },
    "trace": [Function],
    "type": "LogLevel",
    "warn": [Function],
  },
  "logLevel": "silent",
  "logTime": false,
  "port": Object {
    "client": 6000,
    "server": 7000,
  },
  "reload": true,
  "send": Object {
    "errors": true,
    "warnings": true,
  },
  "server": null,
  "stats": Object {
    "context": "<PROJECT_ROOT>",
  },
  "test": false,
  "validTargets": Array [
    "web",
  ],
  "webSocket": Object {
    "host": "localhost",
    "port": 6000,
  },
}
`;

exports[`api options sanity check 1`] = `
Object {
  "host": "127.0.0.1",
  "port": 56130,
}
`;

exports[`api options sanity check 2`] = `
Object {
  "allEntries": false,
  "autoConfigure": true,
  "hmr": true,
  "host": Object {
    "client": "localhost",
    "server": "localhost",
  },
  "https": undefined,
  "log": LogLevel {
    "currentLevel": 5,
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "log": [Function],
    "methodFactory": PrefixFactory {
      "options": Object {
        "level": [Function],
        "name": [Function],
        "template": "{{level}} [90mï½¢{{name}}ï½£[39m: ",
        "time": [Function],
      },
      Symbol(a log instance): [Circular],
      Symbol(valid log levels): Object {
        "DEBUG": 1,
        "ERROR": 4,
        "INFO": 2,
        "SILENT": 5,
        "TRACE": 0,
        "WARN": 3,
      },
    },
    "name": "hot",
    "options": Object {
      "factory": null,
      "level": "silent",
      "name": "hot",
      "prefix": Object {
        "level": [Function],
        "template": "{{level}} [90mï½¢{{name}}ï½£[39m: ",
      },
      "timestamp": false,
      "unique": true,
    },
    "trace": [Function],
    "type": "LogLevel",
    "warn": [Function],
  },
  "logLevel": "silent",
  "logTime": false,
  "port": Object {
    "client": 0,
    "server": 0,
  },
  "reload": true,
  "send": Object {
    "errors": true,
    "warnings": true,
  },
  "server": null,
  "stats": Object {
    "context": "<PROJECT_ROOT>",
  },
  "test": false,
  "validTargets": Array [
    "web",
  ],
  "webSocket": Object {
    "host": "localhost",
    "port": 56130,
  },
}
`;
